digital_ocean_droplet: rebuild=yes + #52128

mz0 opened the pull request:
let droplet rebuild from same or new image

##### SUMMARY
please see #51656.
Fixes #52128.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### ADDITIONAL INFORMATION
I've also reworked internal method 'ensure_power_on' to not GET DO API too early, use locked=False instead of status=active, and renamed it 'ready'.

mz0 committed:
digital_ocean_droplet: state=pristine (#51656)

let droplet rebuild from same or new image

mz0 committed:
state=pristine (#51656) fix doc (add example etc.)

chopraaa commented:
Are we really doing `pristine`? Isn't `rebuild: True` a better option for this?

mz0 commented:
> Are we really doing `pristine`? Isn't `rebuild: True` a better option for this?

are there any precedents of this `rebuild: yes` use in other modules?

BondAnthony commented:
I would echo @chopraaa comment about creating an additional parameter to drive this interaction.

chopraaa commented:
@mz0 General rule of thumb to keep the `state` values consistent across all modules I think.

> I'm not all happy with my proposal - it works only when droplet already exists. Will consider this option.

That's fine, you can just check if the droplet exists when `rebuild: True` is specified. Then call your rebuild function.

mz0 commented:
> @mz0 General rule of thumb is to keep the `state` values consistent across all modules I think.

(vultr_server)[ansible/ansible/blob/e761ea3f9/lib/ansible/modules/cloud/vultr/vultr_server.py#L856] has 
`state=dict(choices=['present', 'absent', 'restarted', 'reinstalled', 'started', 'stopped'], default='present')`
"stopped" or "poweroff" were also on my mind when I considered 'state=pristine', though I've yet to find uses for "stopped".
 
> Also, is there a reason for changing `ensure_power_on` to `ensure_active`? `power_on` is very specific but `active` isn't.
This check is for `'status': 'active'`. 'power_on' is an arbitrary interpretation. 

Moreover I suspect `'locked': 'false'` (which I had to add for rebuild) may be more universal criteria. This needs checking.

mz0 commented:
@chopraaa 
> General rule of thumb is to keep the `state` values consistent across all modules I think.

[vultr_server](https://github.com/ansible/ansible/blob/e761ea3f9/lib/ansible/modules/cloud/vultr/vultr_server.py#L856)
has `state=dict(choices=['present', 'absent', 'restarted', 'reinstalled', 'started', 'stopped'], default='present')`. I had `state=poweroff` (stopped) on my mind when I was considering state=pristine, however I've yet to find any use for `state=poweroff`. Anyway I'll proceed with rebuild=yes.

> Also, is there a reason for changing `ensure_power_on` to `ensure_active`? `power_on` is very specific but `active` isn't.

This is a check for `"status": "active"`. "power_on" is an arbitrary interpretation.

Moreover, `"locked": false` which I had to add for rebuild, may turn more useful. This state:active did not serve me well on new droplets (it's a bit premature) and totally misled on rebuilding.

mz0 committed:
implement rebuild:true and refactor

- implement rebuild: true (re-wire rebuild(..) method)
- deprecate I(unique_name), issue warning.
- move params checks from required_if=([('state', 'present',
    ['name', 'size', 'image', 'region']),]) to module logic.
- add size parameter default, since that's the most obvious thing
  one wants, if desired size can be upgraded later. This default
  is not declared as such so we do the right thing on rebuild.
- kill temp vars, eliminating confusion over names
  and function parameter mutations.
- fix comments
- drop unused import
- style: one function call per line.
- in error messages show droplet identifier in use.
- warn when parameters make no sense (e.g. both id and name supplied)
- rename internal methods:
  create -> get -- common entry point for create/find/rebuild,
  get_by_name -> find_by_name,
  get_droplet -> find_droplet,
  get_addresses -> expose_addresses (that's a mutator. not getter).
- drop trivial core() by moving its body to main().
- doc: correct I(private_networking) description
- doc: update output samples
- doc: remove unneeded 'required: false'
- doc: re NYC2, AMS2, and SFO1 limited availability
  2019.02.07 https://www.digitalocean.com/docs/droplets/overview/
- doc: edit style (e.g. for brevity)

mz0 committed:
RETURN new style YAML, etc.

- better error messages
- NB params description containing  ':' needs quoting!
  (or use '>' trick)

mz0 committed:
doc: add option value types, etc.

mz0 committed:
add debug ver1 (status=active) w.i.p.

mz0 committed:
rework wait logic; add WARNINGS, I(timeout)

I(timeout) done wrong, need reuse digital_ocean_argument_spec. FIXME
issue WARNINGS when droplet exists but given params differ
wait: yes now always returns IP and waits for locked: False.
This way create() and rebuild() may use the same checks.

mz0 committed:
use digital_ocean_argument_spec, WIP

mz0 committed:
better docs, messages. WIP

mz0 committed:
better RETURN - fix types

mz0 committed:
wait logic fixed, same_tags() added, _build() extracted, debug off

mz0 committed:
tidy-up

rebuild show wait before status query (as doc says)
remove unused env_fallback (since digital_ocean_argument_spec re-use)
remove unneeded DOCUMENTATION param 'required: False'
rename back wait_maximum -> wait_timeout (remove alias)
rename wait_minimum -> wait_step, wait_new -> wait_build.
edit docs.

mz0 committed:
debug messages etc.

add conditional (-v) debug messages in 'ops' list
do not fail but warn when no IPv4 found on timeout
wait max(wait_build, wait_timeout) when wait_build=true
expose(): do not copy 'data',
  be explicit about private_ipv4_address,
  do not try adding 'private_ipv6_address' (never seen in the wild ;),
  expose self.ops too if enabled (-v).

mz0 commented:
@gurch101 @BondAnthony @chopraaa @kbreit Please see my latest version (and accompanying doc rendition referenced in #52128) and comment. Please note that the only backward incompatible change I propose: deprecating 'unique_name' param. My version works as if it is always True. The only downside of this - one cannot make multiple 'mydroplet'-named droplets. I hope this is not a loss for anyone of you.

kbreit commented on lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:
```diff
@@ -15,193 +15,351 @@
     description:
-     - array of SSH key (numeric) ID that you would like to be added to the server.
-    required: False
+     - 'List of DO registered SSH key numeric IDs or fingerprints to put in ~root/authorized_keys on creation, e.g.
```
Rename DO to DigitalOcean to avoid acronyms

kbreit commented on lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:
```diff
@@ -15,193 +15,351 @@
-     - This is the slug of the region you would like your server to be created in.
+     - Datacenter slug you would like your droplet to be created in, e.g. C(sfo2), C(sfo1), or C(sgp1).
+     - Required when I(state=present) and the droplet does not yet exist.
+     - "New DO users be aware: due to limited capacity, C(nyc2), c(ams2), and C(sfo1) are
```
Rename DO to DigitalOcean to avoid acronyms

kbreit commented on lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py:
```diff
@@ -15,193 +15,351 @@
+    rebuild: yes
+    image: debian-9-x64  # may be omitted.
+    wait_step: 3.5
+    wait_build: 16  # wait 16 seconds, then GET DO API droplet status with 3.5 seconds pauses, unless default wait_timeout (120 s) is reached.
```
Rename DO to DigitalOcean to avoid acronyms

mz0 replied:
3 DO occurrences replaced as you suggested.

mz0 committed:
replace DO acronym with DigitalOcean etc.

mz0 committed:
fix DOCUMENTATION markup

mz0 committed:
undo checking for 'oauth_token'

allow env. var fallback

mz0 committed:
cleanup params aliases docs

mz0 committed:
return host_ids on state=absent etc.

- when a droplet is deleted it's a good practice to purge knowm_hosts
from that host's key fingerprints. One may loop over host_ids for this.
- add doc for host_ids and msg, identify droplet in msg.
- do not issue false warnings (in rebuild or droplet look-up) when size,
region or image are None.
- cache droplet.id
- consistently use names in debug code

Akasurde commented:
Hi @mz0, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.digitalocean<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
